### DISCUSSION TITLE
Order of edge features in "sparse" mode

### DISCUSSION THREAD

**Discussion by felix-clark** [danielegrattarola/spektral#260] [Q&A]
The documentation suggests that edge features (used by `ECCConv`, for instance) can be in sparse format to avoid padding large arrays with zeros. However, attempting to construct a `Graph` with `e` set to a scipy `coo_matrix` results in an error being raised by the following line:

https://github.com/danielegrattarola/spektral/blob/b6c3795cd12879c44e9591bd04b5d27acb511aab/spektral/data/graph.py#L81

It seems like only dense arrays are permitted for the edge features, unlike for the adjacency matrix (which also allows the type check to pass if `sp.isspmatrix(a)` is `True`). Is this intended, is there a workaround, or am I missing something?

**Comment by danielegrattarola** [comment]
The wording of the documentation can be confusing. Edge features can indeed be in sparse _format_, but that doesn't mean that they are `scipy.sparse` matrices. 
With "sparse format" we indicate the fact that the edge attributes are not tensors of shape `(n_nodes, n_nodes, n_features)`, but instead they are stored in a matrix of shape `(n_edges, n_features)`. This is "sparse", because it lets you keep in memory only the edge features associated with nonzero edges. In a `scipy.sparse` matrix, this is also how sparsity is implemented: by only storing relevant data in np.arrays (e.g., check the `data` field of a `scipy.sparse` matrix).

you can read more about this here: https://graphneural.network/data-modes/

**Comment by felix-clark** [comment]
Thanks for the clarification. The line that threw me off was the following:
"`E` has shape `[edges, e_feat]` so that `E[i]` corresponds to the edge in `A[i // nodes, i % nodes]`."
Based on the indexing of the adjacency matrix in this line I thought that `i` should be in the range `[0, nodes*nodes)`, and thus that `edges` indicated the number of *possible* edges (i.e. `nodes*nodes`). Am I correct in understanding that this line in the documentation doesn't hold (except maybe for fully-connected graphs)?

Can you confirm then that the order of the edge features is taken to be the same ordering as `a.data`, `a.row`, and `a.col` for adjacency matrix `a`? (I just want to make absolutely sure because this is obviously critical.) Doesn't this depend on the representation of the scipy sparse matrix (e.g. `coo_matrix` vs. `csr_matrix`, etc.)? If there is any change in representation under the hood, could this not mix up the association of features with particular edges? Is there anything to be aware of in this regard?

One theoretical footgun is that running the adjacency matrix through a filter could break the correspondence between edge features and the associated edges. There should be no filter applied for `ECCConv`, but I wonder if this would be true in general.

**Comment by felix-clark** [comment]
To try to answer my own question a little bit, I believe that the the data loaders call `spektral.layers.ops.sparse.sp_matrix_to_sp_tensor()` on the adjacency matrix. This function calls `scipy.sparse.find()` to get the row indices, column indices, and data, which itself formats the adjacency as a `coo_matrix` and sums any data in the same place. Then, `tensorflow.sparse.reorder()` is called on the constructed SparseTensor, which puts the indices in standard row-major order (i.e. ordered first by increasing row index, then within each row by increasing column index) if they aren't already.

I don't have a complete understanding of all the tensor operations in `ECCConv.call()`, but I assume that the edge features of the graph must be also be in row-major order for anything in there to make sense. If this understanding is correct, perhaps the documentation could be updated to clarify this? Some users might construct their adjacency matrix in another order and assume that they should also construct their edge features in the same order, which would associate the edge features with the wrong edges.

**Comment by danielegrattarola** [comment]
You are correct, the documentation is imprecise in this regard and this is likely to cause issues to end-users. I will take some time today to:

1. Improve the data loaders so that the specific ordering of the adjacency matrix/edge feature is not something that the users need to worry about
2. Improve the documentation to clarify the meaning of "sparse format".

Thanks for bringing this to my attention.

**Comment by danielegrattarola** [comment]
Ok so point 1 could be a little trickier than expected, but I've gone and added a significant section to the Getting Started tutorial that explains in detail how edge attributes should be represented in a `Graph` object. I've also fixed the wording of the Data Modes tutorial. I'll update the online version of the docs soon.

I also made sure that, if edge attributes are sorted correctly, then all data loaders and layers will know how to deal with them. 
This still requires the users to correctly reorder the edge attributes, but I don't see an easy way out for now. 

I'll change this issue into a discussion so that it remains part of the knowledge base of the library. 

Thanks again
